src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationTxSource.cs

```diff
@@ -0,0 +1,120 @@
+//  Copyright (c) 2021 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Nethermind.AccountAbstraction.Data;
+using Nethermind.AccountAbstraction.Executor;
+using Nethermind.Consensus.Transactions;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Specs;
+using Nethermind.Int256;
+using Nethermind.JsonRpc;
+using Nethermind.Logging;
+
+namespace Nethermind.AccountAbstraction.Source
+{
+    public class UserOperationTxSource : ITxSource
+    {
+        private readonly ILogger _logger;
+        private readonly ISpecProvider _specProvider;
+        private readonly IUserOperationPool _userOperationPool;
+        private readonly IUserOperationSimulator _userOperationSimulator;
+
+        public UserOperationTxSource(
+            IUserOperationPool userOperationPool,
+            IUserOperationSimulator userOperationSimulator,
+            ISpecProvider specProvider,
+            ILogger logger)
+        {
+            _userOperationPool = userOperationPool;
+            _userOperationSimulator = userOperationSimulator;
+            _specProvider = specProvider;
+            _logger = logger;
+        }
+
+        public IEnumerable<Transaction> GetTransactions(BlockHeader parent, long gasLimit)
+        {
+            IDictionary<Address, HashSet<UInt256>> usedAccessList = new Dictionary<Address, HashSet<UInt256>>();
+            IList<UserOperation> userOperationsToInclude = new List<UserOperation>();
+            ulong gasUsed = 0;
+
+            IEnumerable<UserOperation> userOperations =
+                _userOperationPool
+                    .GetUserOperations()
+                    .Where(op => op.MaxFeePerGas >= parent.BaseFeePerGas)
+                    .OrderByDescending(op => CalculateUserOperationPremiumGasPrice(op, parent.BaseFeePerGas));
```

Comment:
Can't those come already sorted from the pool? (maybe not)
See how we sort in TxPoolTxSource, maybe that will be usefull.

Reply:
@marcindsobczak do you know if they come sorted? i'm not sure

Reply:
In UserOperationPool they are sorted by MaxPriorityFeePerGas. In UserOperationTxSource we need them sorted by effective priority fee. In TxPoolTxSource we are resorting txs as well, but using IEnumerable, so with some optimization.

There is a way to sort it properly in UserOperationPool. We will need to change comparer in UserOperationPool from comparing MaxPriorityFeePerGas to comparing effective priority fee. Then we need to recalculate effective priority fee with every base fee change (so every block) and in UserOperationTxSource we will be able to just use the order from pool. In this scenario we need to add logic for this recalculations and test it properly before merging to the master. Not sure if it's worth it for now.

Reply:
Skip it for now